### PR TITLE
Assignment fourteen asynchronous programming

### PR DESCRIPTION
**Pull request by Sruthi-Subaraja**
ASSIGNMENT 14 -ASYNCHRONOUS PROGRAMMING
To gain in-depth understanding and practical experience of Async/Await, Task Parallel Library (TPL), and Multi-Threading in C#.
TASK 1 - Understanding and implementing async/await.
TASK 2 - Implementing and understanding Task Parallel Library.
TASK 3 - Advanced understanding and usage of multi-threading.
TASK 4 - Implementing multi-layered async/await operations in a real-world scenario.
TASK 5 - Debugging and fixing deadlock conditions.
TASK 6 - Real-world application of ConfigureAwait in a console application with thread tracking.
TASK 7 - Understanding the difference between async void and async task with exceptions.

### CODE REVIEW

File: src/AsynchronousProgramming/Deadlock.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        /// <summary>
+        /// A method to perform some operation and cause deadlock by holding the main thread.
+        /// </summary>
+        /// <returns>A task object</returns>
+        public static async Task DeadlockMethod()
+        {
+            var result = await SomeAsyncOperation();
+
+            // var result = SomeAsyncOperation().Result; // will make the main thread wait causing a deadlock
```
Comment (Tharun-Sundar) [inline src/AsynchronousProgramming/Deadlock.cs:20]: Explain why blocking the main thread causes deadlock. Doesn't using await make the thread wait until the awaited operation is done?

File: src/AsynchronousProgramming/ApplicationWithConfigureAwait.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        }
+
+        /// <summary>
+        /// Calculates square of array elements.
+        /// </summary>
+        /// <returns>A task object of int type</returns>
+        public static async Task<int> MethodB()
+        {
+            var result = await MethodA().ConfigureAwait(false);
+            TaskParallelLibraryImplementation.PrintArrayUsingParallerThreads(result);
```
Comment (Tharun-Sundar) [inline src/AsynchronousProgramming/ApplicationWithConfigureAwait.cs:39]: Why is this method called here?

File: tests/unit/AsynchronousProgramming/AsyncAndAwaitClassTest/AsyncAndAwaitClassTest.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+namespace AsyncAndAwaitClassTest
+{
+    using AsynchronousProgramming;
+    public class AsyncAndAwaitClassTest
+    {
+        [Fact]
+        public async void URL_GettingContentsFromURL_ReturnTheContentsOfURL()
+        {
+            string url = "https://learn.microsoft.com/en-us/dotnet/api/system.threading.thread?view=net-7.0";
```
Comment (Tharun-Sundar) [inline tests/unit/AsynchronousProgramming/AsyncAndAwaitClassTest/AsyncAndAwaitClassTest.cs:9]: This test has the same logic used in the method which is wring. Tests should not have logic like this. It should only call methods written in the main code.

File: tests/unit/AsynchronousProgramming/AsyncAndAwaitClassTest/AsyncAndAwaitClassTest.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
+namespace AsyncAndAwaitClassTest
+{
+    using AsynchronousProgramming;
+    public class AsyncAndAwaitClassTest
+    {
+        [Fact]
+        public async void URL_GettingContentsFromURL_ReturnTheContentsOfURL()
```
Comment (Tharun-Sundar) [inline tests/unit/AsynchronousProgramming/AsyncAndAwaitClassTest/AsyncAndAwaitClassTest.cs:7]:
```suggestion
        public async void Url_GettingContentsFromUrl_ReturnTheContentsOfUrl()
```

**Review by Tharun-Sundar — CHANGES_REQUESTED** [review]

File: src/AsynchronousProgramming/AsyncAndAwait.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    /// To understand the concept of async and await.
+    /// </summary>
+    public class AsyncAndAwait
+    {
+        /// <summary>
+        /// Get the contents of a URL.
+        /// </summary>
+        /// <param name="url">URL for which content is extracted.</param>
+        /// <returns>A task object of type string.</returns>
+        public static async Task<string> GetContentsOfURL(string url)
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AsynchronousProgramming/AsyncAndAwait.cs:17]: Done

[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AsynchronousProgramming/AsyncAndAwait.cs:17]: Okay

File: src/AsynchronousProgramming/Program.cs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+                switch (operationToPerform)
+                {
+                    case 1:
+                        string url = "https://learn.microsoft.com/en-us/dotnet/api/system.threading.thread?view=net-7.0";
+                        var content = await AsyncAndAwait.GetContentsOfURL(url);
+                        Console.WriteLine("The contents of the URL is " + content);
+                        break;
+                    case 2:
+                        int[] arrayOfIntegers = new int[10000];
+                        for (int i = 0; i < 10000; i++)
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AsynchronousProgramming/Program.cs:42]: Done

File: src/AsynchronousProgramming/Program.cs [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+                        {
+                            ExceptionHandling.VoidMethod();
+                        }
+                        catch (Exception ex)
+                        {
+                            Console.WriteLine(ex.Message);
+                        }
+
+                        break;
+                    default:
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AsynchronousProgramming/Program.cs:86]: Missed that. I updated it now.

File: src/AsynchronousProgramming/TaskParallelLibraryImplementation.cs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    /// <summary>
+    /// Calculate square of array elements.
+    /// </summary>
+    public class TaskParallelLibraryImplementation
+    {
+        /// <summary>
+        /// Square and print an array of integers.
+        /// </summary>
+        /// <param name="arrayOfIntegers">The array of integer that is squared.</param>
+        public static void PrintArrayUsingParallerThreads(int[] arrayOfIntegers)
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AsynchronousProgramming/TaskParallelLibraryImplementation.cs:18]: Changed it.

File: src/AsynchronousProgramming/ApplicationWithConfigureAwait.cs (hunk 2 above)
Comment (Sruthi-Subaraja) [reply src/AsynchronousProgramming/ApplicationWithConfigureAwait.cs:39]: It is not necessary, I only added that to have a long running operation. I removed it and only have MethodA() as method that executes a long running operation.

File: src/AsynchronousProgramming/Deadlock.cs (hunk 1 above)
Comment (Sruthi-Subaraja) [reply src/AsynchronousProgramming/Deadlock.cs:20]: I'll explain it in the README.md.

File: src/AsynchronousProgramming/MultiThreading.cs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    /// <summary>
+    /// Using multi threading to perform basic calculations.
+    /// </summary>
+    public class MultiThreading
+    {
+        /// <summary>
+        /// Perform addition and multiplication using multi threading.
+        /// </summary>
+        /// <returns>A task object</returns>
+        public static async Task UsingMultiThreadingToAddAndMultiply()
```
[the remark this answers is not in the document]
Comment (Sruthi-Subaraja) [reply src/AsynchronousProgramming/MultiThreading.cs:16]: The methods will be executed by the caller thread if the methods are called without a new thread. The thread.Join() could be used after starting both the threads which would indicate that the methods are started the execution simultaneously but will wait until both the execution is called.

File: tests/unit/AsynchronousProgramming/AsyncAndAwaitClassTest/AsyncAndAwaitClassTest.cs (hunk 4 above)
Comment (Sruthi-Subaraja) [reply tests/unit/AsynchronousProgramming/AsyncAndAwaitClassTest/AsyncAndAwaitClassTest.cs:7]: Done.

File: tests/unit/AsynchronousProgramming/AsyncAndAwaitClassTest/AsyncAndAwaitClassTest.cs (hunk 3 above)
Comment (Sruthi-Subaraja) [reply tests/unit/AsynchronousProgramming/AsyncAndAwaitClassTest/AsyncAndAwaitClassTest.cs:9]: The GetStringAsync() method is not overridable, how else can I mock and test the method?

Comment (Tharun-Sundar) [reply tests/unit/AsynchronousProgramming/AsyncAndAwaitClassTest/AsyncAndAwaitClassTest.cs:9]: Then this test is not needed

Comment (Sruthi-Subaraja) [reply tests/unit/AsynchronousProgramming/AsyncAndAwaitClassTest/AsyncAndAwaitClassTest.cs:9]: Okay.

**Review by Tharun-Sundar — APPROVED** [review]

**Pull request merged**